File: Old Code/WurmRancher/WurmRancher/Level.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.Windows.Media.Imaging;
using System.IO.IsolatedStorage;

namespace WurmRacher
{
    public abstract class Level
    {        
        bool gameover = false;
        protected static IsolatedStorageSettings levelData = IsolatedStorageSettings.ApplicationSettings;

        public const string ApplicationKey = "868a65a194464d31b4ca21f8977bd01a";
        public const string ApplicationSecret = "wrs";

        public enum CompletionStatusEnum
        {
            Unattempted, Attempted, Completed
        }

        public CompletionStatusEnum CompletionStatus = CompletionStatusEnum.Unattempted;


        //This returns true if the game is over.
        public bool GameOver
        {
            get { return gameover; }
        }

        public virtual int WeedGrowthRate{
            get { return Timing.RelativeTimeToFrames(3); } 
        }
        public virtual int PoisonWeedGrowthRate
        {
            get { return Timing.RelativeTimeToFrames(4); }
        }
        public virtual string Name
        {
            get {return "unknown";}
        }
        public virtual string Description
        {
            get {return "no description available";}
        }
        public Theme Theme
        {
            get { return theme; }
        }
        public virtual bool NoUserControl
        {
            get { return false; }
        }
        //public virtual string HighScore
        //{
        //    get { return "--"; }
        //}
        public string ID
        {
            get { return Name + Version; }
        }
        protected virtual string Version
        {
            get { return "1"; }
        }
        //protected virtual void LoadHighScore()
        //{
        //}
        public virtual string QuickObjectives
        {
            get { return "Objectives unknown."; }
        }
        public virtual bool SeedDisabled
        {
            get { return false; }
        }
        //public virtual void ResetHighScore() { }

        public virtual string HighScoreName
        {
            get { return "none"; }
        }
        public virtual bool HasHighScore
        {
            get { return true; }
        }
        public virtual bool MakeFeedersAtWill
        {
            get { return false; }
        }

        Theme theme;
        public Level(Theme theme_)
        {
            theme = theme_;
            if (levelData.Contains(ID + "_c"))
            {
                this.CompletionStatus = CompletionStatusEnum.Completed;
                //this.LoadHighScore();
            }
            else if (levelData.Contains(ID + "_a"))
                this.CompletionStatus = CompletionStatusEnum.Attempted;         

        }

        public event EventHandler StatusChanged;

        
        protected int elapsed_frames=0;
        public virtual void Update(GameControl theControl)
        {
            if (gameover)
                return;
            elapsed_frames++;
            if (elapsed_frames == 1 && !this.NoUserControl)
                theControl.ShowMessage(Description, Name);
        }
        public virtual void InitializeLevel(GameControl theControl)
        {
            elapsed_frames = 0;
            gameover = false;
            if (CompletionStatus == CompletionStatusEnum.Unattempted)
            {
                CompletionStatus = CompletionStatusEnum.Attempted;
                levelData[this.ID + "_a"] = true;
            }
            if (!gameover && StatusChanged != null)
                StatusChanged(this, new EventArgs());
        }

        protected virtual void Victory(GameControl theControl, string message = null, _3XH.IHighScoreCtrl highScoreCtr = null)
        {
            if (gameover)
                return;
            if (message == null)
                message = "You have completed the mission " + Name + "!";
            CompletionStatus = CompletionStatusEnum.Completed;
            levelData[ID + "_c"] = true;
            gameover = true;
            if (StatusChanged != null)
                StatusChanged(this, new EventArgs());
            theControl.ShowVictory(message, highScoreCtr);

        }

        protected virtual void Defeat(GameControl theControl, string message = null)
        {
            if (gameover)
                return;
            if (message == null)
                message = "You failed to complete the mission objectives.";            
            gameover = true;
            theControl.ShowDefeat(message);
            if (StatusChanged != null)
                StatusChanged(this, new EventArgs());
        }

        protected bool IntervalTimeIsUp(double rel_interval, double rel_offset = 0)
        {
            int interval = Timing.RelativeTimeToFrames(rel_interval);
            int offset = Timing.RelativeTimeToFrames(rel_offset);
            return (elapsed_frames % interval == offset);
        }

        protected bool OneTimeTriggerIsUp(double rel_time)
        {
            return (elapsed_frames == Timing.RelativeTimeToFrames(rel_time));
        }

        //theControl should call this if the level is stopped prematurely.
        public void Quit()
        {
            if (!gameover && StatusChanged != null)
                StatusChanged(this, new EventArgs());
        }
    }
}

File: ts/level.ts
import { ReportDefeat, ReportVictory } from "./gameControl.js";
import { Theme } from "./theme.js";
import { createCookie, readCookie } from "./cookies.js";

const cookie_expire_length = 30;

export enum CompletionStatus{
    Unattempted = 0,
    Attempted = 1,
    Completed = 2
}

export class Level{
    completionStatus : CompletionStatus;
    gameover : boolean;
    theme : Theme;

    //default values
    WeedGrowthRate = 3000;
    PoisonWeedGrowthRate = 4000;
    Name = "unknown";
    Description = "no description available";
    NoUserControl = false;
    Version = "1";
    QuickObjectives = "objectives unknown";
    SeedDisabled = false;
    MakeFeedersAtWill = false;
    elapsed_time : number;
    last_time_step :number;

    score : number;
    low_score_best = true;

    high_score : number = null;
    
    get ID():string{
        return this.Name + this.Version;
    }

    constructor(theme : Theme){
        this.theme = theme;
        // load saved data?
        let hss = readCookie(this.highScoreName);
        if (hss)
            this.high_score  = Number(readCookie(this.highScoreName));
    }

    Update(time_step: number) : void{
        if (this.gameover)
            return;
        this.last_time_step = time_step;
        this.elapsed_time += time_step;
    
        //this.elapsed_frames++;
        //if (this.elapsed_frames == 1 && !this.NoUserControl)
        //    ShowMessage(this.Name + ": " + this.Description );
    }

    InitializeLevel() :void{
        this.elapsed_time = 0;
        this.gameover = false;
        if (this.completionStatus == CompletionStatus.Unattempted)
        {
            this.completionStatus = CompletionStatus.Attempted;
            //levelData[this.ID + "_a"] = true;
        }
        //if (!this.gameover && StatusChanged != null)
        //    StatusChanged(this, new EventArgs());
    }

    Victory(message :string = null) : void{
        if (this.gameover)
            return;
        if (message == null)
            message = "You have completed the mission " + this.Name + "!";
        this.completionStatus = CompletionStatus.Completed;
        //levelData[this.ID + "_c"] = true;
        this.gameover = true;
        //if (StatusChanged != null)
        //    StatusChanged(this, new EventArgs());
        if ((this.low_score_best && this.score < this.high_score) || (!this.low_score_best && this.score > this.high_score))
            createCookie(this.highScoreName,this.score, cookie_expire_length);
        ReportVictory(message);
    }

    Defeat(message : string = null) : void{
        if (this.gameover)
            return;
        if (message == null)
            message = "You failed to complete the mission objectives.";            
        this.gameover = true;
        ReportDefeat(message);
        //if (StatusChanged != null)
        //    StatusChanged(this, new EventArgs());
    }

    

    //theControl should call this if the level is stopped prematurely.
    Quit() : void
    {
        //if (!this.gameover && StatusChanged != null)
        //    StatusChanged(this, new EventArgs());
    }

    IntervalTimeIsUp( interval : number, offset : number=0) : boolean{
        return ((this.elapsed_time + offset) % interval < this.last_time_step);
        //let interval = RelativeTimeToFrames(rel_interval);
        //let offset = RelativeTimeToFrames(real_offset);
        //console.log(this.elapsed_frames, interval,offset, this.elapsed_frames % interval == offset)
        //return (this.elapsed_frames % interval == offset); // make sure this works ok with numbers!!!
    }

    OneTimeTriggerIsUp(rel_time:number){
        return ((this.elapsed_time >= rel_time) &&  this.elapsed_time - rel_time < this.last_time_step);
    }

    get highScoreName():string{
        return this.Name.replace(/\s+/g, '');
    }

}
